src/persist-client/src/internal/trace.rs

```diff
@@ -935,15 +992,318 @@ impl<T: Timestamp + Lattice + Codec64> SpineBatch<T> {
         }
 
         if let Some((id, range)) = self.find_replacement_range(res) {
-            self.perform_subset_replacement(res, id, range)
+            self.perform_subset_replacement(&res.output, id, range, None)
         } else {
             ApplyMergeResult::NotAppliedNoMatch
         }
     }
 
+    #[allow(dead_code)]
+    fn construct_batch_with_runs_replaced(
+        original: &HollowBatch<T>,
+        run_ids: &[RunId],
+        replacement: &HollowBatch<T>,
+    ) -> Result<HollowBatch<T>, ApplyMergeResult> {
+        if run_ids.is_empty() {
+            return Err(ApplyMergeResult::NotAppliedNoMatch);
+        }
+
+        assert!(
+            replacement.run_meta.len() <= 1,
+            "replacement must have exactly 0 or 1 runs"
+        );
+
+        let run_ids = run_ids.to_vec();
+
+        let start_id = run_ids[0];
+        let end_id = *run_ids.last().unwrap();
+
+        // 0. Find the indices of the runs in the original batch.
```

Comment:
I've suggested the possibility before, but now I'm sure of it: this logic is too complex for what is effectively a batch build over `original.runs().filter(|(meta, _)| !run_ids.contains(id).chain(replacement.runs())`. I encourage you to simplify it! Using the higher-level methods like `.runs` instead of a bunch of vec indexing ought to help, or https://github.com/MaterializeInc/materialize/pull/32955 if you want to take a look at it.

(It also imposes a bunch of restrictions that aren't documented but would be fine if they were necessary (like the single-run restriction on our input, or that runs are contiguous) but AFAICT they're only there to not make the logic below more complicated, and I think you can probably get rid of them with a more straightforward impl.)